File: baboon-compiler/src/main/resources/baboon-runtime/scala/BaboonConversions.scala
package baboon.runtime.shared {

  import scala.collection.mutable
  import scala.reflect.ClassTag

  trait Conversion {
    def typeFrom: Class[?]
    def versionFrom: String

    def typeTo: Class[?]
    def versionTo: String

    def typeId: String
  }

  trait BaboonGeneratedConversion extends Conversion {
    def convert[C](context: Option[C], conversions: AbstractBaboonConversions, from: BaboonGenerated): BaboonGenerated
  }

  abstract class AbstractConversion[From, To](
    implicit fromTag: ClassTag[From],
    toTag: ClassTag[To],
  ) extends BaboonGeneratedConversion {
    protected def doConvert[Ctx](context: Option[Ctx], conversions: AbstractBaboonConversions, from: From): To

    private def validateBaboonType(obj: Any): Unit = {
      obj match {
        case bg: BaboonGenerated =>
          val conversionTypeIsExactType = typeId == bg.baboonTypeIdentifier
          bg match {
            case bga: BaboonAdtMemberMeta =>
              val conversionTypeIsAdtType = typeId == bga.baboonAdtTypeIdentifier
              if (!conversionTypeIsAdtType && !conversionTypeIsExactType) {
                throw new IllegalArgumentException(
                  s"Provided instance is adt=${bga.baboonAdtTypeIdentifier} exact=${bg.baboonTypeIdentifier}, one of which must be $typeId"
                )
              }
            case _ =>
              if (!conversionTypeIsExactType) {
                throw new IllegalArgumentException(
                  s"Provided instance is ${bg.baboonTypeIdentifier} but must be $typeId"
                )
              }
          }
        case _ =>
      }
    }

    def convert[Ctx](context: Option[Ctx], conversions: AbstractBaboonConversions, from: From): To = {
      validateBaboonType(from)
      val result = doConvert(context, conversions, from)
      validateBaboonType(result)
      result
    }

    override def convert[Ctx](context: Option[Ctx], conversions: AbstractBaboonConversions, from: BaboonGenerated): BaboonGenerated = {
      from match {
        case f: From =>
          val res = convert[Ctx](context, conversions, f)
          res match {
            case bg: BaboonGenerated => bg
            case _ =>
              throw new IllegalArgumentException(
                s"Can't use IBaboonGeneratedConversion interface for non IBaboonGenerated return type To = ${typeTo.getName}"
              )
          }

        case _ =>
          throw new Exception(
            s"Can't use IBaboonGeneratedConversion interface when 'from' is not of type ${typeFrom.getName}"
          )
      }
    }

    override final def typeFrom: Class[?] = fromTag.runtimeClass
    override final def typeTo: Class[?]   = toTag.runtimeClass
  }

  case class ConversionKey(from: Class[?], to: Class[?])

  trait AbstractBaboonConversions {
    private val conversions     = mutable.Map.empty[ConversionKey, Conversion]
    private val typeConversions = mutable.Map.empty[Class[?], List[Conversion]]

    def register[F, T](
      conversion: AbstractConversion[F, T]
    )(implicit
      tagF: ClassTag[F],
      tagT: ClassTag[T],
    ): Unit = {
      val fromClass       = tagF.runtimeClass
      val toClass         = tagT.runtimeClass
      val key             = ConversionKey(fromClass, toClass)
      val fromConversions = typeConversions.getOrElse(fromClass, Nil)
      conversions.update(key, conversion)
      typeConversions.update(fromClass, fromConversions.appended(conversion))
    }

    def convertWithContext[C](ctx: Option[C], from: BaboonGenerated, conversion: Conversion): BaboonGenerated = {
      conversion.asInstanceOf[BaboonGeneratedConversion].convert(ctx, this, from)
    }

    def convert(from: BaboonGenerated, conversion: Conversion): BaboonGenerated = {
      conversion.asInstanceOf[BaboonGeneratedConversion].convert(None, this, from)
    }

    def convertWithContext[C, F, T](
      context: Option[C],
      from: F,
    )(implicit
      tagF: ClassTag[F],
      tagT: ClassTag[T],
    ): T = {
      val key = ConversionKey(tagF.runtimeClass, tagT.runtimeClass)
      conversions(key).asInstanceOf[AbstractConversion[F, T]].convert[C](context, this, from)
    }

    def convert[F <: BaboonGenerated, T <: BaboonGenerated](
      from: F
    )(implicit
      tagF: ClassTag[F],
      tagT: ClassTag[T],
    ): T = {
      convertWithContext[Any, F, T](None, from)(tagF, tagT)
    }

    def findConversions(value: BaboonGenerated): List[Conversion] = {
      typeConversions.get(implicitly[ClassTag[BaboonGenerated]].runtimeClass) match {
        case Some(value) => value
        case None =>
          value match {
            case m: BaboonAdtMemberMeta => typeConversions.getOrElse(m.baboonAdtType, Nil)
            case _                      => Nil
          }
      }
    }

    def versionsFrom: List[String]
    def versionTo: String
  }
}

File: baboon-compiler/src/main/resources/baboon-runtime/kotlin/BaboonConversions.kt
package baboon.runtime.shared

interface Conversion {
    val typeFrom: Class<*>
    val versionFrom: String
    val typeTo: Class<*>
    val versionTo: String
    val typeId: String
}

interface BaboonGeneratedConversion : Conversion {
    fun <C> convert(context: C?, conversions: AbstractBaboonConversions, from: BaboonGenerated): BaboonGenerated
}

abstract class AbstractConversion<From : Any, To : Any>(
    private val fromClass: Class<From>,
    private val toClass: Class<To>,
) : BaboonGeneratedConversion {
    protected abstract fun <Ctx> doConvert(context: Ctx?, conversions: AbstractBaboonConversions, from: From): To

    private fun validateBaboonType(obj: Any) {
        when (obj) {
            is BaboonGenerated -> {
                val conversionTypeIsExactType = typeId == obj.baboonTypeIdentifier
                when (obj) {
                    is BaboonAdtMemberMeta -> {
                        val conversionTypeIsAdtType = typeId == obj.baboonAdtTypeIdentifier
                        if (!conversionTypeIsAdtType && !conversionTypeIsExactType) {
                            throw IllegalArgumentException(
                                "Provided instance is adt=${obj.baboonAdtTypeIdentifier} exact=${obj.baboonTypeIdentifier}, one of which must be $typeId"
                            )
                        }
                    }
                    else -> {
                        if (!conversionTypeIsExactType) {
                            throw IllegalArgumentException(
                                "Provided instance is ${obj.baboonTypeIdentifier} but must be $typeId"
                            )
                        }
                    }
                }
            }
        }
    }

    fun <Ctx> convert(context: Ctx?, conversions: AbstractBaboonConversions, from: From): To {
        validateBaboonType(from)
        val result = doConvert(context, conversions, from)
        validateBaboonType(result)
        return result
    }

    @Suppress("UNCHECKED_CAST")
    override fun <C> convert(context: C?, conversions: AbstractBaboonConversions, from: BaboonGenerated): BaboonGenerated {
        val f = fromClass.cast(from) as From
        val res = convert(context, conversions, f)
        return res as BaboonGenerated
    }

    override val typeFrom: Class<*> get() = fromClass
    override val typeTo: Class<*> get() = toClass
}

data class ConversionKey(val from: Class<*>, val to: Class<*>)

abstract class AbstractBaboonConversions {
    private val conversions = mutableMapOf<ConversionKey, Conversion>()
    private val typeConversions = mutableMapOf<Class<*>, MutableList<Conversion>>()

    fun <F : Any, T : Any> register(conversion: AbstractConversion<F, T>) {
        val fromClass = conversion.typeFrom
        val toClass = conversion.typeTo
        val key = ConversionKey(fromClass, toClass)
        conversions[key] = conversion
        typeConversions.getOrPut(fromClass) { mutableListOf() }.add(conversion)
    }

    fun <C> convertWithContext(ctx: C?, from: BaboonGenerated, conversion: Conversion): BaboonGenerated {
        @Suppress("UNCHECKED_CAST")
        return (conversion as BaboonGeneratedConversion).convert(ctx, this, from)
    }

    fun <C> convertWithContext(ctx: C?, from: BaboonGenerated, fromClass: Class<*>, toClass: Class<*>): BaboonGenerated {
        val key = ConversionKey(fromClass, toClass)
        val conv = conversions[key] ?: throw NoSuchElementException("No conversion from $fromClass to $toClass")
        @Suppress("UNCHECKED_CAST")
        return (conv as BaboonGeneratedConversion).convert(ctx, this, from)
    }

    fun convert(from: BaboonGenerated, conversion: Conversion): BaboonGenerated {
        return convertWithContext<Any>(null, from, conversion)
    }

    fun findConversions(value: BaboonGenerated): List<Conversion> {
        val fromType = value::class.java
        return typeConversions[fromType]
            ?: when (value) {
                is BaboonAdtMemberMeta -> typeConversions[value.baboonAdtType] ?: emptyList()
                else -> emptyList()
            }
    }

    abstract val versionsFrom: List<String>
    abstract val versionTo: String
}
